### ISSUE TITLE
Project idea discussion thread

### ISSUE THREAD

**Issue by Krzysztof-Cieslak** [fsprojects/atom-fsharp#1]
This repository has been created as a place for discussion and future development of rich F# experience on top of [Atom](https://atom.io/) editor. Inspired a lot by [OmniSharp](https://github.com/OmniSharp/omnisharp-atom) team, I strongly believe it is possible to create very powerful, cross platform, and lightweight F# code editor giving extensibility of Atom editor.

The code already in repository is just a simple proof of concept for communication between Atom package and [F# Compiler Services](https://github.com/fsharp/FSharp.Compiler.Service) - current iteration is using [fsautocomplete.exe](https://github.com/fsharp/fsharpbinding/tree/master/FSharp.AutoComplete) tool. As for now it does nothing, running plugin ( ctrl + alt + o) just run set of commands against predefined F# files and prints output to console. Also current version is using CoffeScript which might not be the best choice.

Before starting serious implementation few things should be decided and i hope for some input.
Below are some topics and possible solutions to them ( for sure there are more ;)). Feel free to give Your opinion on all or any of them. 

*  The architecture of system
   - One big plugin
   - Set of smaller composable packages
*  The language of choice
   - Plain JavaScript
   - Any Compile2JS languages ( CoffeeScript / TypeScript ?)
   - Using F# and [FunScript](https://github.com/ZachBray/FunScript) to compile to JS
* The way of comunication with F# Compiler Services
   - FSharp.Autocomplete
   - Building Suave based local WebService hosting F# Compiler Services
   - Using [edge.js](https://github.com/tjanczuk/edge) to call F# Compiler Services directly ( probably small bridge / adapter library necessary for best experiance)

Also would want to know what is most important, what should be scope of project at start, and what fancy features would be welcomed ( Maybe support for [Paket](https://github.com/fsprojects/Paket) and [Fake](https://github.com/fsharp/FAKE) from the editor level? It should be easy to add)

**Comment by jwood803** [comment]
I'm guessing running the F# Interactive in Atom would be part of it, too?

**Comment by reidev275** [comment]
I'd like to see smaller components. Could always do a package of packages for an install all sort of thing. 

Using FunScript could allow us to dog food it which could be cool. 

Agree with @jwoood803 that the interactive would be great

**Comment by dsyme** [comment]
Why not add F# support to Omnisharp?

**Comment by Krzysztof-Cieslak** [comment]
@dsyme, If anywhere I would prefer to merge it with our F# bindings project - they don't have Atom support at all, this idea is in the scope they are working on, and probably this project will be using some libraries created by them. It should be another discussion with guys maintaining F# binding if it would be worth to try merge whole F# bindings with OmniSharp and I don't feel I'm competent enough to start it right now. 

 @jwood803, For sure Interactive should be part of it

Personally, I would also go for set of smaller components. As @reidev275 suggested it should be easy to create installer for all parts.

FunScript was an idea suggested by@tpetricek in our twitter conversation. It seems great although I personally don't have a lot experience with it and would love to hear also others opinions.

**Comment by dsyme** [comment]
I should have mentioned that F# + Atom would be fantastic to have - it would find all sorts of uses in practice.  For example we would likely use it in a future revised http://tryfsharp.org (to replace the HTML5 editor currently used there which came from an internal Microsoft project)

**Comment by Krzysztof-Cieslak** [comment]
I have implemented basic Autocomplete. For sure it's just basic hacky implementation but it is good proof of concept. Still plugin is implemented using CoffeeScript - I don't have enough experience in FunScript. And here goes gif ;)

![fsharpatom1](https://cloud.githubusercontent.com/assets/5427083/6740089/aa0fd6e2-ce7d-11e4-9b51-40bcdd7a7ef7.gif)

**Comment by jwood803** [comment]
That's awesome! I'm liking how this looks already.

**Comment by Krzysztof-Cieslak** [comment]
Well, It still has some problems:
  * Atom has idea of grammars ( definition of language syntax highlighting) and as for now https://github.com/jbtule/language-fsharp is used. As far as I know @tpetricek was [investigating building dynamic grammar using F# Compiler Services](https://discuss.atom.io/t/syntax-highlighting-using-existing-tokenizer/13608)
  * As for now it's working just for standalone .fs files or .fsx files with #load and #r directives. It doesn't support fsproj yet.
  * It's written in CoffeeScript ;) And personally I'm not a big fan of this language

**Comment by dsyme** [comment]
The usual thing with fsharpbinding is when a new editor binding comes along we make one or two people associated with that editor binding administrators on the project.  That's what happened with Emacs for example.

**Comment by Krzysztof-Cieslak** [comment]
@dsyme, Well seems good for me. If we manage to make here something useful I would be more than happy to merge it with fsharpbinding. 

As for now I would just like to get thing rolling. Just waiting for some more input from community. And Tomas' opinions since he was interested, he has some cool ideas, he investigated some topics earlier, and conversation with him pushed me to start this project

**Comment by tpetricek** [comment]
I'd love to get involved in this :-). I think I should have some time on Sunday/Monday to have a look. I think doing the early experiments here is probably a good idea. If it turns into something useful, we can move it to fsharpbindings!

**Comment by dsyme** [comment]
It may be that moving to fsharpbindings is not needed - when that was first started there weren't many F# open projects. The F# community is now much more adept at making and sustaining new projects.

However I'd recommend moving to a github org like fsprojects (or similar) sooner or later.

**Comment by Krzysztof-Cieslak** [comment]
Heh, we have our amazing http://fhash.org also as a github org, we could put it there ;) [just kidding]
Again, don't see any problems with moving it to some F# github org.

I have added sample error / warning highlighting implementation.
This time just a picture.

![Error Highlighting](https://cloud.githubusercontent.com/assets/5427083/6754248/b038b694-cf1b-11e4-8e0e-91bc10db3af2.png)

**Comment by Krzysztof-Cieslak** [comment]
Getting to know Atom API continues - this time bottom panel with errors / warnings.

![errorpanel](https://cloud.githubusercontent.com/assets/5427083/6761103/d80508e2-cf4f-11e4-8515-4d41ae0baf28.png)

**Comment by tpetricek** [comment]
I managed to create a very simple demo using FunScript: https://github.com/tpetricek/FunScript.Atom

It actually looks pretty nice! If we manage to generate FunScript bindings for the atom package from `atom.d.ts`, then it should be able to do all we need. The next question if we can get F# Data to work with FunScript again and use that to parse the JSON coming from the console app! 

That said, as much as I'd love to use F# for writing the F# plugin, the current story really is that if you want to use it, you need to kind of know both F# and JavaScript (and probably especially with Atom). So, just writing everything using ordinary CoffeeScript might be easier way to get it done....

**Comment by Krzysztof-Cieslak** [comment]
I have pushed proof of concept for autocomplete created using F# and FunScript. Still not all features from CoffeScript implementation are ported but it should be done Soon™

For me development was OK and I think that advantages of F# outshines possible drawbacks ( such as bit hacky and hard-coded code generation and some other FunScript problems). Also creating tooling in F# can again be selling point not only for this project but also for F# as whole ( I would for sure mention in every F# talk I give ;) ). And people most interested in development of F# support for Atom are F# developers - so it will make easier for people to contribute. So even if I can agree that implementing it in CoffeeScript may be easier I still think it's worth doing it in F#.

Tomas, we should definitely try to do something with `atom.d.ts` as it would make things even easier. What's exact problem there? When I tried to generate bindings I got no unhandled exceptions as in cases of previous errors I worked around. Error log file was also empty.

**Comment by Krzysztof-Cieslak** [comment]
I have ported last part of CoffeScript implementation to F#. There are still few problems - declaration of view is as for now very naive, I still don't use atom bindings provided by Tomas in https://github.com/Krzysztof-Cieslak/FSharp.Atom/pull/2 - integrating it will be last step for this proof of concept.

**Comment by 7sharp9** [comment]
I use Atom to edit F# files some of the time, it would be nice to see semantic formatting and completion available :-)

**Comment by 7sharp9** [comment]
@Krzysztof-Cieslak @tpetricek IS this buildable on non Windows?

**Comment by 7sharp9** [comment]
@Krzysztof-Cieslak 
Re: **`Using edge.js to call F# Compiler Services directly ( probably small bridge / adapter library necessary for best experience)`**
You could probably use `edge.fs` too, an all F# async path could be funky :-)

**Comment by tpetricek** [comment]
@7sharp9 I did all the work on Mac with Xamarin :-). 

I think doing the calls via the console (or small HTTP server) is fine. The only thing that's not asynchronous in Atom is colorization - and that we can probably solve later (though it would be nice to use the F# compiler for that too - even in Xamarin, the colorization is really easy to break).

**Comment by 7sharp9** [comment]
@tpetricek The colourisation in XS is based on two parts, a regex parser that highlights the basic syntax, and a semantic formatter that is based on the AST being available to apply some further colourisation to members, values, fields, closures etc.

As for the Compiler doing all the work, I think this might be out of scope, I see the compiler providing the capability to get all the information to do the further colourisation, like what we do in the F# adding in XS.  This is mainly done via the FSharpSymbol API thats been steadily taking shape over the last year.  There are already Symbol version of most core compiler functions rather that the older text based results that you were forced to crack apart.

**Comment by Krzysztof-Cieslak** [comment]
@tpetricek
Re: `By "all the work" I mean the little thing I contributed.
Obviously @Krzysztof-Cieslak did all the work!` That explains why most of this code is so bad ;)

@7sharp9 
Thanks for pointing out `edge.js` and giving some insight about XS. As for now plugin is using static colorization created by @jbtule -  https://github.com/jbtule/language-fsharp . Actually from point of view of autocomplete and error list I just check if Grammar for currently active TextEditor has name "F#", so it's not coupled to any colorization implementation. 

So, my idea right now:
* Already existing code is base for *core* plugin - it will cover such things as autocomplete, error highlighting, error list, tooltips - all those features are in scope of FSharp.Autocomplete and we can communicate with it using standard I/O
* We should make some investigation about creating better syntax highlighter which will replace or complement static highlighting we are using right now.
 * Creating few smaller plugins for set of tools which are commonly used in F# community (such as Paket support, FAKE runner etc) and other useful features (refactoring, navigation  etc)

**Comment by jwood803** [comment]
@Krzysztof-Cieslak Is this all still to be done using CoffeeScript?

**Comment by Krzysztof-Cieslak** [comment]
@jwood803, I don't think CoffeScript is necessary, F# story is OK (especially after adding generated bindings for atom), only drawback for now is that it sometimes requires updating `atom.d.ts` file and regenerating bindings since original `atom.d.ts` is bit outdated.
However every plugin is separate entity and can be developed using different stack ( tho, i don't see any reason not to use F# as for now ;) )

**Comment by jwood803** [comment]
That's awesome! I hope to help contribute in that case :boom: 

That CoffeeScript was a little intimidating, to tell the truth. :p

**Comment by Krzysztof-Cieslak** [comment]
@jwood803, Well, I've just started using it since that's what Atom community does - all examples, tutorials and documentation is using it. As i said before - I'm not big fan of CoffeeScript, if F# was not an option I would rather go to TypeScript or plain JavaScript. Fortunately ( mostly thanks to Tomas' work) we can use F# without many problems.

**Comment by tpetricek** [comment]
@7sharp9 I might be missing something - but I think the basic colorization should be pretty straightforward using the [F# Compiler Service's tokenizer](http://fsharp.github.io/FSharp.Compiler.Service/tokenizer.html). I use this to colorize things in F# Formatting and it is pretty fast & correct. So I think the regex colorization could be replaced using this. 

(And the same is the case for Atom, if we can figure out how to make synchronous calls, or if we can compile the F# tokenizer into JavasScript to embed it directly :-))

**Comment by 7sharp9** [comment]
I use the tokenizer in XS too, the regex is fallback for when the AST or lexing is broken.  We use the symbols to allow colorisation of AST dependent stuff like computation expressions, field declaration, field usage etc etc.

**Comment by tpetricek** [comment]
Hmm that is odd, Xamarin never worked correctly for me with e.g. """ strings or nested comments - even when my code was all correct. The lexer should not need correct F# source though (and it certainly does not need AST that can be parsed).

Is this something that has changed recently? Perhaps I just need to install the latest version...

**Comment by 7sharp9** [comment]
Oh I know the lexer doesn't need an AST, semantic highlighting does though.   Field usage, field declaration have different highlight colours.   When the AST is returned async, the semantic part is applied over the basic highlighting.  

I would like to improve the resilience though.  

5.9 should be out soon which has fixes in that area.

**Comment by Krzysztof-Cieslak** [comment]
@7sharp9, @tpetricek I won't even pretend I understand what you guys are talking about ;) Can we move discussion about syntax highlighting to separate issue? 
I was also planning to create separate discussion threads for core plugin and Paket  plugin ( which very basic version I should push tomorrow so maybe someone interested in contributing will have easy starting  point)

**Comment by 7sharp9** [comment]
@Krzysztof-Cieslak Yeah Ill shut up :-)  @tpetricek We can discuss this more at https://github.com/fsharp/FSharpMDXS/issues/6

**Comment by Krzysztof-Cieslak** [comment]
@7sharp9, I don't see any problems with this discussion as such, just wanted to create separate discussions for different parts of plugin :).

**Comment by Ciantic** [comment]
@Krzysztof-Cieslak Cany you answer to this: https://github.com/fsharp/fsharpbinding/issues/990 

I don't know what are the implications of moving

**Comment by mcgee** [comment]
Great project! I would like to contribute, but I don't know where to begin. I have several ideas how I could help, but I'm not sure that my opinion on what should be done coincides with community. Do you mind if I formalize the ideas as issues so we can discuss them separately? Most of them are about methodology, i.e.:
* First of all, every open source project needs documentation to help new contributors start development. I already created small page "How to setup development environment", but it should be more detailed, I think. It also would be great to describe how this plugin work, its main parts and what external components are used.
*  What about automated testing? Should contributors write unit-tests and what framework will be used for that?
* Small notes about project organization:
  * Is it necessary to store ``.dll`` files in ``lib`` folder under version control? Should it be installed via NuGet instead?
  * The same thought could be said about ``fsautocomplete.exe`` and its dlls
  * Should plugin ``core`` be renamed to something less general?
  * Should  ``core.fs`` be broken to several files to make navigation easier?

...and so on

**Comment by Krzysztof-Cieslak** [comment]
@mcgee, Those are all good points! As for now, everything seems bit chaotic but I'm planning to do some work on those "formal" things when decisions discussed in fsharp/fsharpbinding#990 will be made.
Just to answer some of questions:  

* Testing is bit tricky, I'm not sure how to handle it from FunScript point of view, while testing generated JS code also can be difficult (since FunScript is generating some really messy code) ( /cc: @alfonsogarciacaro)
* `DLL` files from `lib` are nowhere on nuget - those are files generated by me ( and there is option to re-generate them in our FAKE build file)
* For  `fsautocomplete.exe` we need to wait for changes in organization of projects discussed in fsharp/fsharpbinding#990. 
* `core` for sure should and will be renamed ;)
* splitting `core.fs` into several files seems like a good idea.

**Comment by paulyoung** [comment]
I wasn't aware of this project and started a dynamic grammar based on https://github.com/TypeStrong/atom-typescript/blob/master/CONTRIBUTING.md#dynamic-grammar

I currently only have a package that sends lines to FSharp.Compiler.Service for tokenization. This still needs a bit of work but I'd be happy to contribute my efforts back here or something more official.

My intention was to use that as research for when I get access to [Nuclide](http://nuclide.io/) so that a lot of the work there can be leveraged.

***

Nuclide is a set of Atom packages being developed by Facebook.

My post requesting more info about: adding more languages: https://facebook.com/groups/833105363425649?view=permalink&id=853799101356275

**Comment by Krzysztof-Cieslak** [comment]
Hey @paulyoung.

We would love to have dynamic grammar for syntax highlighting and will be very happy with any contributions. I know @tpetricek was interested mostly in this topic and also did some investigation on it.

**Comment by 7sharp9** [comment]
@paulyoung what is the dynamic grammar, is it like semantic highlighting?

**Comment by paulyoung** [comment]
@7sharp9 instead of using the [TextMate grammar format](https://manual.macromates.com/en/language_grammars) for syntax highlighting it sends lines to `FSharp.Compiler.Service` to [tokenize lines](http://fsharp.github.io/FSharp.Compiler.Service/tokenizer.html).

My initial motivation was mostly because I hate the old plist format and find that it's a pain to keep in sync as languages change.

I'm using a similar approach with a language we've been developing at work (where syntax has been changing rapidly) and wondered if doing the same for FSharp would be fast enough and open doors to using other features like the [editor services](http://fsharp.github.io/FSharp.Compiler.Service/editor.html).

**Comment by tpetricek** [comment]
@paulyoung This would be amazing! The current grammar is flaky in various places and I think actually calling the compiler service is the only correct way of fixing that.

I asked about how to do this at the Atom forums some time ago:

 - I got [a couple of useful pointers in response](https://discuss.atom.io/t/syntax-highlighting-using-existing-tokenizer/13608)
 - And there is [another related thread I found](https://discuss.atom.io/t/defining-a-language-grammar-via-code/14784)

The tricky thing seems to be that Atom does this synchronously - but AFAIK, we only have asynchronous ways for calling the F# Compiler service (but perhaps there is some way to hack that?). Another option would be to get FunScript or WebSharper to compile the lexer part of F# Compiler Service into JavaScript (which would be doable, but might be non-trivial amount of work...)

**Comment by paulyoung** [comment]
Here's my [question and findings](https://discuss.atom.io/t/language-features-driven-by-something-other-than-textmate-style-grammar/16586).

I think support for async tokenization is being tracked here: https://github.com/atom/first-mate/issues/34

Includes interest from someone at OmniSharp, among others.

**Comment by paulyoung** [comment]
> get FunScript or WebSharper to compile the lexer part of F# Compiler Service into JavaScript 

This might be more performant and easier to bundle.

**Comment by 7sharp9** [comment]
AFAIK the tokenization in FCS is synchronous, only retrieving symbols to assist would be async.

**Comment by 7sharp9** [comment]
i.e
```fsharp
let tokens = 
                    [| let state = ref 0L
                      for line in alllines do
                        let lineTxt = editor.GetLineText(line)
                        let tokenizer = sourceTok.CreateLineTokenizer(lineTxt)
                        let rec parseLine() =
                            [ match tokenizer.ScanToken(!state) with
                              | Some(tok), nstate ->
                                  yield tok
                                  state := nstate
                                  yield! parseLine()
                              | None, nstate -> state := nstate ]
                        yield line, parseLine() |]
```

**Comment by 7sharp9** [comment]
It can be time consuming to call though as the state of the previous line is required, otherwise information such as an active `#if` will be lost etc

**Comment by paulyoung** [comment]
It's definitely looking like it's going to be too slow from what I could see last night.

**Comment by 7sharp9** [comment]
Tokenisation only, without any async symbol retrial should be around 30-50ms for normal sized files.

**Comment by paulyoung** [comment]
Admittedly my focus was on mapping tokens to what Atom expects. I had hoped that some optimization could come later.

**Comment by Krzysztof-Cieslak** [comment]
Discussions about project to be continued in #96
